test/lib.sh

```diff
@@ -59,3 +59,59 @@ EOF
 hr() {
     echo "========================="
 }
+
+diff() {
+	for file in `ls` ; do 
+		if [ -d $file ] ; then 
+			pushd $file  > /dev/null
+			diff $1/$file 
+			popd  > /dev/null
+		elif [ -f $file ] ; then
+			cmp $file $1/$file
+			if [ ! $? = 0 ] ; then
+				rc=`expr $rc+1`
+			fi
+		fi
+	done
+	return 0
+}
+
+
+compare() {
+	if [ $1 = "_data" ] ; then
+		if [ ! -f $repo/boilerplate/_data/last_boilerplate_commit ] ; then
+			rc=`expr $rc+1`
+		fi
+	else
+		pushd $1  > /dev/null
+		diff $REPO_ROOT/boilerplate/$1
+		popd > /dev/null
+	fi
+	return 0
+}
+
+check_update() {
+	declare -i rc
+	
+	pushd $repo/boilerplate > /dev/null
+	
+	compare _data
+	compare _lib
+	
+	while read convention ; do
+	  if [ -d $REPO_ROOT/boilerplate/$convention ] ; then
+		  compare $convention
+	  fi
+	done < $repo/boilerplate/update.cfg
+	
+	popd > /dev/null
+	
+	return $rc
+}
+
+add_convention() {
+	if [ $(grep $1\$ $repo/boilerplate/update.cfg | wc -l) == 0 ] ; then
+		echo $1 >> $repo/boilerplate/update.cfg
+	fi
+	make update_boilerplate
```

Comment:
We should avoid running `make update_boilerplate` in these helper methods, instead letting the calling environment do it explicitly. For example, at some point we'll want to write a test that adds multiple conventions at once before updating.